state/storage.go

```diff
@@ -306,6 +306,77 @@ func (s *Storage) MetadataForAllRooms(txn *sqlx.Tx, tempTableName string, result
 	return nil
 }
 
+// ResetMetadataState updates the given metadata in-place to reflect the current state
+// of the room.
+// TODO: could be brittle. Might be better to just create a new RoomMetadata from scratch
+// and copy over the fields we don't want to reload.
+func (s *Storage) ResetMetadataState(metadata *internal.RoomMetadata) error {
+	var events []Event
+	err := s.DB.Select(&events, `
+	WITH snapshot(events, membership_events) AS (
+        SELECT events, membership_events
+        FROM syncv3_snapshots
+            JOIN syncv3_rooms ON snapshot_id = current_snapshot_id
+        WHERE syncv3_rooms.room_id = $1
+    )
+	SELECT event_id, event_type, state_key, event, membership
+	FROM syncv3_events JOIN snapshot ON (
+		event_nid = ANY (ARRAY_CAT(events, membership_events))
+	)
+	WHERE (event_type IN ('m.room.name', 'm.room.avatar', 'm.room.canonical_alias') AND state_key = '')
+	   OR (event_type = 'm.room.member' AND membership IN ('join', '_join', 'invite', '_invite'))
+	ORDER BY event_nid ASC
+	;`, metadata.RoomID)
```

Comment:
How long does this take on m.org?

Reply:
I did some testing on Matrix HQ---let me rerun and get you some numbers.

Reply:
According to notes this was ~200ms for Matrix HQ. Running it now is ~270ms.

```
postgres=> \set room_id !OGEhHVWSdvArJzumhm:matrix.org \set snapshot_id 682264 \\
postgres=> 
postgres=> EXPLAIN ANALYZE
postgres-> WITH snapshot(events, membership_events) AS (
postgres(>     SELECT events, membership_events
postgres(>     FROM syncv3_snapshots
postgres(>         JOIN syncv3_rooms ON snapshot_id = current_snapshot_id
postgres(>     WHERE syncv3_rooms.room_id = :'room_id'
postgres(> )
postgres-> SELECT event_id, event_type, state_key, event, membership
postgres-> FROM syncv3_events JOIN snapshot ON (
postgres(>         event_nid = ANY (ARRAY_CAT(events, membership_events))
postgres(>     )
postgres-> WHERE (event_type IN ('m.room.name', 'm.room.avatar', 'm.room.canonical_alias') AND state_key = '')
postgres->    OR (event_type = 'm.room.member' AND membership IN ('join', '_join', 'invite', '_invite'))
postgres-> ORDER BY event_nid ASC
postgres-> ;
                                                                                                                   QUERY PLAN                 
                                                                                                   
----------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------
 Sort  (cost=24.28..24.28 rows=2 width=501) (actual time=246.204..258.618 rows=48046 loops=1)
   Sort Key: syncv3_events.event_nid
   Sort Method: external merge  Disk: 19976kB
   ->  Nested Loop  (cost=1.15..24.27 rows=2 width=501) (actual time=17.091..200.713 rows=48046 loops=1)
         ->  Nested Loop  (cost=0.71..16.75 rows=1 width=271) (actual time=0.032..0.034 rows=1 loops=1)
               ->  Index Scan using syncv3_rooms_pkey on syncv3_rooms  (cost=0.29..8.30 rows=1 width=8) (actual time=0.020..0.021 rows=1 loops
=1)
                     Index Cond: (room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text)
               ->  Index Scan using syncv3_snapshots_snapshot_id_room_id_key on syncv3_snapshots  (cost=0.42..8.44 rows=1 width=279) (actual t
ime=0.010..0.010 rows=1 loops=1)
                     Index Cond: (snapshot_id = syncv3_rooms.current_snapshot_id)
         ->  Index Scan using syncv3_events_pkey on syncv3_events  (cost=0.44..7.50 rows=2 width=501) (actual time=15.542..192.561 rows=48046 
loops=1)
               Index Cond: (event_nid = ANY (array_cat(syncv3_snapshots.events, syncv3_snapshots.membership_events)))
               Filter: (((event_type = ANY ('{m.room.name,m.room.avatar,m.room.canonical_alias}'::text[])) AND (state_key = ''::text)) OR ((ev
ent_type = 'm.room.member'::text) AND (membership = ANY ('{join,_join,invite,_invite}'::text[]))))
               Rows Removed by Filter: 53701
 Planning Time: 2.699 ms
 Execution Time: 268.729 ms
(15 rows)
```

Reply:
Eyeballing it, the ORDER BY adds ~70ms, which surprised me! (I want to process the Nids in ascending order to pick out the heroes.)

Reply:
I wonder if it might be quicker to do the sorting in Go? Suggest this is good enough to leave as-is for now though.

(Also I'm surprised that we use the 6 newest heroes, not the 6 oldest)